### PR TITLE
Creating DiscoveryProviderFactory

### PR DESCRIPTION
**Pull request by garrettjonesgoogle**
Net removal of more than 500 lines :-)

* Also changing the result type of GapicProvider - now returning Map<String, Doc> so that the CodeGeneratorApi class can just call ToolUtil.writeFiles itself.

### CODE REVIEW

File: src/main/java/com/google/api/codegen/GeneratorBuilderUtil.java [hunk 1]
```diff
@@ -14,13 +14,6 @@
  */
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/GeneratorBuilderUtil.java]: The class doc comments here are no longer accurate. (They reference a "codegen provider".)

File: src/main/java/com/google/api/codegen/gapic/GapicProvider.java [hunk 2]
```diff
@@ -30,13 +31,13 @@
...
-   * Runs code generation and puts the output in outputPath.
+   * Runs code generation and returns a map relative file paths to generated Doc.
    */
-  void generate(String outputPath) throws Exception;
+  Map<String, Doc> generate();
 
   /**
-   * Generates code for a single snippet. Returns a map from service interface to code for the service.
-   * Returns null if generation failed.
+   * Runs code generation for a single snippet and returns a map relative file
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/gapic/GapicProvider.java]: "map relative" ---> "map from relative"

File: src/main/java/com/google/api/codegen/GeneratorBuilderUtil.java (hunk 1 above)
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/GeneratorBuilderUtil.java]: I'll take care of that, and additionally I think I'll rename the class while I'm at it.

File: src/main/java/com/google/api/codegen/gapic/GapicProvider.java [hunk 3]
```diff
@@ -30,13 +31,13 @@
   List<String> getSnippetFileNames();
 
   /**
-   * Runs code generation and puts the output in outputPath.
+   * Runs code generation and returns a map relative file paths to generated Doc.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/gapic/GapicProvider.java]: "map relative" ---> "map from relative"

**Comment by geigerj** [comment]
Hooray for net removal of code! Looks good to me other than a few comments.

**Comment by garrettjonesgoogle** [comment]
I have addressed your comments, PTAL.

**Comment by pongad** [comment]
LGTM if test passes.

**Comment by geigerj** [comment]
LGTM

**Pull request merged**